src/scatter.jl

```diff
@@ -0,0 +1,179 @@
+export scatter!, scatter
+
+## Scatter API
+#   - Scatter:
+#     - scatter(op, src, idx, dims)
+#     - scatter!(op, dst, src, idx, dims)
+#   - Scatter destination backpropagation
+#     - ∇scatter_dst!
+#   - Scatter source backpropagation
+#     - ∇scatter_src
+#     - ∇scatter_src!
+#
+
+_check_input(idx::AbstractArray{<:Integer}, arr) = checkbounds(arr, minimum(idx):maximum(idx))
+
+function _check_input(idx::AbstractArray{<:Tuple}, arr)
+    pairs = map(xs -> Base.OneTo(maximum(xs)), zip(idx...))
+    checkbounds(arr, pairs...)
+end
+
+function _check_output(idx::AbstractArray{<:IntOrTuple}, dst, src, dims)
+    idx_dims = size(idx)
+    dst_dims = size(dst)
+    src_dims = size(src)
+    dst_dims[1:dims] == src_dims[1:dims] || throw(ArgumentError("dst and src must have the same dimensions in the first $(dims) dimensions"))
+end
+
+
+"""
+    scatter!(op, dst, src, idx, dims)
+
+Scatter operation, which scatters data in `src` and assigns to `dst` according to `idx`.
+With the data going to the same place, specified operation is applied on to reduce data.
+For each index `k` in `idx`, accumulate values in `dst` according to
+
+    dst[idx[k]...] = (op).(dst[idx[k]...], src[k...])
+
+# Arguments
+- `op`: operations to be applied on `dst` and `src`, e.g. `+`, `-`, `*`, `/`, `max` and `min`.
+- `dst`: the destination for `src` to aggregate to. This argument will be mutated.
+- `src`: the source data for aggregating.
+- `idx`: the mapping for aggregation from source (index) to destination (value).
+The index of `idx` is corresponding to the index of `src` and the value of `idx` is
+corresponding to the index of `dst`. The value of `idx` can be `Int` or `Tuple` type.
+- `dims`: the number of dimensions to be view as an operational unit from the beginning.
+If dims=0 is given, operational unit is scalar. Default dims is 1.
+
+Examples for dims are lists here:
+
+- `dims=0`: dst[idx[k]...] = op(dst[idx[k]...], src[k...])
+- `dims=1`: dst[:, idx[k]...] = (op).(dst[:, idx[k]...], src[:, k...])
+- `dims=2`: dst[:, :, idx[k]...] = (op).(dst[:, :, idx[k]...], src[:, :, k...])
+"""
+function scatter!(op, dst::AbstractArray{T}, src::AbstractArray{T}, idx::AbstractArray{<:IntOrTuple};
+                  dims::Integer=1) where {T<:Real}
```

Comment:
unless I'm missing something, I think we can remove `dims` entirely because essentially it is determined by the other arrays' shapes and eltypes, right? Look at how this is done in #280

Reply:
I spend couples of days checking the generality of current API is good. I have new commit pushed.